### ISSUE TITLE
Redux devtools not working Next 13

### ISSUE THREAD

**Issue by fasetto** [vercel/next.js#43690]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103
    Binaries:
      Node: 16.15.1
      npm: 8.11.0
      Yarn: 1.22.19
      pnpm: 7.17.1
    Relevant packages:
      next: 13.0.7-canary.1
      eslint-config-next: 13.0.4
      react: 18.2.0
      react-dom: 18.2.0
```

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://stackblitz.com/edit/nextjs-z7lnws?file=app/store.js

### To Reproduce

* Go to the reproduction link. 
* Open the redux dev tools extension. 
* Check if you can see state for `counter` 

### Describe the Bug

Its only show the next-router instance. There is no other instance for me to see the states & events related to my application. 

<img width="437" alt="image" src="https://user-images.githubusercontent.com/20834021/205492993-d4cdc1d6-8b24-4a69-85a6-42355fb7b151.png">

Not sure if its related but, I saw this warning on the console.
```
wait  - compiling...
wait  - compiling /_error (client and server)...
warn  - ./node_modules/.pnpm/react-redux@8.0.5_2zx2umvpluuhvlq44va5bta2da/node_modules/react-redux/es/utils/reactBatchedUpdates.js
export 'unstable_batchedUpdates' (reexported as 'unstable_batchedUpdates') was not found in 'react-dom' (possible exports: __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED, createPortal, flushSync, preinit, preload, version)
```

### Expected Behavior

I should be able to see the instance of my application with related redux states & events.

<sub>[NEXT-1389](https://linear.app/vercel/issue/NEXT-1389/redux-devtools-not-working-next-13)</sub>

**Comment by ciruz** [comment]
Same for me with the ``/app`` folder structure ``redux-toolkit`` and Typescript.

I tried different configurations with ``createStore()`` and using a middleware, or with the simple approach ``configureStore()``, but in the end it looks the same for me. There also other people like this guy https://stackoverflow.com/a/74539739/12237723

**Comment by zkSoju** [comment]
Also running into this issue when using the `/app` folder, seems to be the problem.

**Comment by GautierCo** [comment]
Same problem when I'm using the `/app` folder. (with `redux-toolkit`)

**Comment by mezhinsky** [comment]
Same here, dev tools connection only if you change the init state in reducer

**Comment by angelmtztrc** [comment]
Same here, dev tools only connect after a change, but I can't get the actual or current value of my state in dev tools.

**Comment by switz** [comment]
Seeing the same warning/error:

```
../../node_modules/react-redux/es/utils/reactBatchedUpdates.js
export 'unstable_batchedUpdates' (reexported as 'unstable_batchedUpdates') was not found in 'react-dom' (possible exports: __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED, createPortal, flushSync, preinit, preload, version)
``

**Comment by vishalkumar2806** [comment]
Im also having the same problem

**Comment by vishalkumar2806** [comment]
annyone got any solution

**Comment by markerikson** [comment]
FYI folks, I saw this get reported in the Redux Toolkit repo today, and figured out the issue. Next's internal "router reducer" is trying to connect to the Redux DevTools so that the Next devs can debug it, but the connection code accidentally hardcoded the "instance ID".  That's overriding any Redux DevTools connection set up by your application.

I've notified the Next devs and they should hopefully fix it early next week.

**Comment by vishalkumar2806** [comment]
I reported it.... But if you know the solution please create a file and we will replace it in our source

**Comment by markerikson** [comment]
@vishalkumar2806 : There is no solution yet - it's a bug in Next's code.

**Comment by vishalkumar2806** [comment]
no you said that you give suggestion if you can solve it please do it

**Comment by vishalkumar2806** [comment]
I got a temporary solution for this problem... actually we can't the state because of our instance but its exist... so what you can do is get a copy of state and cosole log it like this
do everything as explained in the tutorial of redux toolkit 
  const state = useSelector( (state) => state )
  const count = useSelector((state : any) => state.counter.value)
  const dispatch = useDispatch()
  console.log(state)
now i can see the current state in console... tho this is temporary like you can preserve state like in redux dev tools

**Comment by NathanVG** [comment]
> FYI folks, I saw this get reported in the Redux Toolkit repo today, and figured out the issue. Next's internal "router reducer" is trying to connect to the Redux DevTools so that the Next devs can debug it, but the connection code accidentally hardcoded the "instance ID". That's overriding any Redux DevTools connection set up by your application.
> 
> I've notified the Next devs and they should hopefully fix it early next week.

Any news on this? I ran into this issue today.

**Comment by AdrianMachado** [comment]
Also running into this

**Comment by codebycarlos** [comment]
Also running into this with React Query https://github.com/TanStack/query/issues/5023

**Comment by ciruz** [comment]
This issue was also discussed here https://github.com/reduxjs/redux-toolkit/issues/3154

>Update: confirmed that this is a Next bug. They tried to wrap useReducer to talk to the Redux DevTools, but hardcoded the instanceId:
>https://github.com/vercel/next.js/blob/0d7f7653e81a4f2aa16be7ead286e09b22761f52/packages/next/src/client/components/use-reducer-with-devtools.ts#LL144-L149C6

> I've passed along a suggestion to have them fix that and call the Redux DevTools connect() method correctly.
> https://github.com/reduxjs/redux-toolkit/issues/3154#issuecomment-1416439233

> @markerikson

There was a fix by @timneutkens https://github.com/vercel/next.js/commit/fea8c8e13a40967e927ce28c4e492b2736e22bce addressing this issue.

For me this issue is gone, I'm on next `13.2.4` right now and the redux devtools work as before and as expected.

**Comment by madi-tracksuit** [comment]
Is there any movement on this? Also running into the issue with `react-query`, seems Next doesn't export `unstable_batchedUpdates` on the server / when running build.

**Comment by Shubhrajyoti-Dey-FrosTiK** [comment]
By any chance if you are using NextJS

Edge runtime causes this problem as apparently it creates its own `dom` where `unstable_batchedUpdates` is not exported

So just remove `edge` runtime from the file and/or from 'next.config.js`

in file generally its written as 

```
export const config = {
    runtime: "edge",
}
```

or 

```
export runtime = "edge"
```


Remove these from the file or replace `edge` with `nodejs`

**Comment by engineering-ks** [comment]
So I cannot use edge runtime in a project that has react-query ? That is so unfortunaate.

**Comment by GabeDuarteM** [comment]
That's probably the same as https://github.com/vercel/next.js/issues/47057#issuecomment-1597807892 I imagine. 
There, I posted an updated minimal reproduction repo using the latest canary and react-query versions, I'll also post it here in case it's helpful:

[GabeDuarteM/nextjs-bug-unstable_batchedUpdates](https://github.com/GabeDuarteM/nextjs-bug-unstable_batchedUpdates)

and yeah, it seems that for now the edge runtime basically breaks anything that depends on `unstable_batchedUpdates`, which includes `react-query` atm, and from what I can see also quite a few other frameworks, like `redux` for example...

The workaround, for now, is not to use the edge runtime but would be great to have this properly fixed, as those are huge and popular libraries and I can imagine this would be a huge dealbreaker for people using edge on Vercel...

**Comment by ztanner** [comment]
Hey everyone -- the issue with Redux dev tools appears to be fixed in #45649 and the issues with `unstable_batchedUpdates` also seem resolved as of https://github.com/facebook/react/pull/27028 (cc @GabeDuarteM your repro worked for me after upgrading Next).

Going to close this out but if you continue to see issues please open a new ticket!
